### PR TITLE
Feature/problem logic

### LINKED ISSUE #6: Implement Basic CRUD Services for Problems
Create a service layer to handle core CRUD operations for the Problem model. This will separate business logic from Django views and make it easier to maintain and test.

Scope:
- Implement services for creating, updating and deleting problems.
- Apply basic validation in the service layer.
- Write unit tests for all service functions.

Acceptance Criteria:
-  In the `apps/problems/services.py` file, add the `create_problem` function with the basic fields.
- Add the `get_all_problems` function to return active problems.
- Add the `get_problem_by_id` function to return a single active problem by id.
- Add the `update_problem` function to handle changes to a problem.
- Add the `deactivate_problem` function, which “safely” deletes a problem 
- Validation in the service layer: a problem cannot be created with a difficulty level not in DIFFICULTY_CHOICES.
- All service functions covered by unit tests (file `tests/test_problem_services.py`).

### LINKED ISSUE #8: Implement TestCase Service Layer
Create a service layer to handle creation of test cases for coding problems. This separates test case business rules from Django views and ensures all test cases are created consistently.

Scope:
- Implement a service for creating a new TestCase.
- Validate test case input before saving.
- Ensure proper linkage between Problem and TestCase.
- Write unit tests for the test case creation flow.

Acceptance Criteria:
- In apps/problems/services.py add the create_test_case function.
- The function must accept problem, input_data, expected_output, and is_hidden as arguments.
- Add the `get_test_case_for_problems` function.
- Add the `get_test_case_by_id` function.
- A test case cannot be created if input_data is empty or if expected_output is empty.
- A new TestCase is correctly linked to the provided Problem.
- If is_hidden is not provided, it defaults to False.
- Unit tests cover successful test case creation and validation failures for empty input data or expected output.

Tasks:
- Create or update apps/problems/services.py.
- Implement create_test_case.
- Add validation for input_data and expected_output.
- Add tests in tests/test_testcase_services.py.

### PR DESCRIPTION
**Pull request by Eveline-projects**
This PR introduces a business logic layer (Service Layer) that separates database operations from views: 

- Service Issue (#6): Full CRUD (create, update, delete) with difficulty level validation and support for partial updates.

- TestCase Service (#8 ): Creating test cases with validation of empty fields (.strip()).

- Tests: A complete set of unit tests in pytest for both services, verifying data correctness and error handling.

Related tasks:

- Closes #6 

- Closes #8

### CODE REVIEW

File: apps/problems/services.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        raise ValidationError('Invalid difficulty level')
+
+    return Problem.objects.create(
+        title=title,
+        description=description,
+        difficulty=difficulty,
+        category=category,
+    )
+
+def update_problem(
```
Comment (Filip28) [inline apps/problems/services.py:19]: **Style:** PEP 8 requires **two** blank lines between top-level functions. There is only one blank line here between `create_problem` and `update_problem`.

File: apps/problems/services.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        problem: Problem,
+        title: str | None = None,
+        description: str | None = None,
+        difficulty: str | None = None,
+        category: str | None = None,
+) -> Problem:
+
```
Comment (Filip28) [inline apps/problems/services.py]: **Style:** Extra blank line after the closing `) -> Problem:` of the function signature. Remove it — the function body should start directly on the next line.

File: apps/problems/services.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        problem.description = description
+    if difficulty is not None:
+        problem.difficulty = difficulty
+    if category is not None:
+        problem.category = category
+
+    problem.save()
+    return problem
+
+def delete_problem(problem: Problem) -> Problem:
```
Comment (Filip28) [inline apps/problems/services.py]: **Naming:** `delete_problem` is misleading — it performs a *soft delete*, not a real deletion. Consider renaming to `deactivate_problem` (or `soft_delete_problem`) to make the behavior clear without needing to look at the implementation.

**Review by Filip28 — COMMENTED** [review]
Good implementation overall. The service layer cleanly separates business logic from views, validation is correct, and test coverage is solid. A few style issues and one missing test case noted inline.

File: apps/problems/services.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    return problem
+
+
+def create_test_case(problem: Problem, input_data: str, expected_output: str, is_hidden: bool = False) -> TestCase:
+    if not input_data.strip():
+        raise ValidationError('Input data cannot be empty')
+    if not expected_output.strip():
+        raise ValidationError('Expected output cannot be empty')
+
+
```
Comment (Filip28) [inline apps/problems/services.py]: **Style:** Double blank line inside a function body. Use a single blank line (or no blank line) between the validation block and the `return` statement.

File: tests/test_problem_services.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    def test_create_problem_should_raise_error_on_invalid_difficulty(self):
+        with pytest.raises(ValidationError):
+            services.create_problem(
+                title='Test',
+                description='Test',
+                difficulty='invalid_level',
+                category='Test'
+            )
+
```
Comment (Filip28) [inline tests/test_problem_services.py:40]: **Missing test:** `update_problem` validates the difficulty level and raises `ValidationError` on an invalid value, but there is no test covering that branch. Consider adding:
```python
def test_update_problem_should_raise_error_on_invalid_difficulty(self, problem: Problem):
    with pytest.raises(ValidationError):
        services.update_problem(problem, difficulty="invalid_level")
```

File: tests/test_problem_services.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    def test_update_problem_should_change_difficulty(self, problem: Problem):
+        services.update_problem(
+            problem,
+            difficulty='hard'
+        )
+        problem.refresh_from_db()
+        assert problem.difficulty == 'hard'
+
+
```
Comment (Filip28) [inline tests/test_problem_services.py]: **Style:** Double blank line inside a class body. PEP 8 recommends a **single** blank line between methods within a class.

File: tests/test_test_case_services.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        with pytest.raises(ValidationError) as excinfo:
+            services.create_test_case(
+                problem=problem,
+                input_data='  ',
+                expected_output='4',
+            )
+
+        assert 'Input data cannot be empty' in str(excinfo.value)
+
+
```
Comment (Filip28) [inline tests/test_test_case_services.py]: **Style:** Double blank line inside a class body. Use a single blank line between methods.

File: tests/test_users.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 
 from apps.users.forms import CreateUserForm, ArchitectLoginForm
 
-
+@pytest.mark.django_db
```
Comment (Filip28) [inline tests/test_users.py:8]: **Nit:** `@pytest.mark.django_db` is unnecessary here. This test only instantiates a form and checks `is_valid()` in memory — no database access happens. The decorator is harmless but adds a small overhead and implies DB usage where there is none.

Comment (Eveline-projects) [reply tests/test_users.py:8]: I removed the `user` fixture because it wasn't actually used directly in the test. However, I had to add `@pytest.mark.django_db` because `CreateUserForm` makes database queries during the `is_valid()` validation. Without this, the test raises a `RuntimeError`.

Comment (Filip28) [reply tests/test_users.py:8]: Ok, good

**Review by Filip28 — APPROVED** [review]
LGTM
